rp2040-hal/src/i2c.rs

```diff
@@ -1,3 +1,384 @@
 //! Inter-Integrated Circuit (I2C) bus
+// Based on: https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_i2c/i2c.c
+// Structure from: https://github.com/japaric/stm32f30x-hal/blob/master/src/i2c.rs
 // See [Chapter 4 Section 3](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
-// TODO
+
+use crate::gpio::pin::bank0::{
+    Gpio0, Gpio1, Gpio10, Gpio11, Gpio12, Gpio13, Gpio14, Gpio15, Gpio16, Gpio17, Gpio18, Gpio19,
+    Gpio2, Gpio20, Gpio21, Gpio26, Gpio27, Gpio3, Gpio4, Gpio5, Gpio6, Gpio7, Gpio8, Gpio9,
+};
+use embedded_time::rate::Hertz;
+use hal::blocking::i2c::{Write, WriteRead};
+use rp2040_pac::{I2C0, I2C1, RESETS};
+
+/// I2C error
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum Error {
+    /// I2c abort with error
+    Abort(u32),
+}
+
+#[doc(hidden)]
+mod sealed {
+
+    use crate::gpio::pin::bank0::{
+        Gpio0, Gpio1, Gpio10, Gpio11, Gpio12, Gpio13, Gpio14, Gpio15, Gpio16, Gpio17, Gpio18,
+        Gpio19, Gpio2, Gpio20, Gpio21, Gpio26, Gpio27, Gpio3, Gpio4, Gpio5, Gpio6, Gpio7, Gpio8,
+        Gpio9,
+    };
+
+    pub trait Sealed {}
```

Comment:
We already have `crate::typelevel::Sealed`, which all of these `PinId`s already implement

Reply:
Nice! Fixed!